### PR TITLE
#173 Add test to ensure no false-positive in PMD UnusedPrivateField

### PR DESCRIPTION
**Pull request by mkordas**
For #173:
* issue described in ticket was false-positive in PMD which was fixed recently
* test is added to make sure issue is fixed and will not reoccur

### CODE REVIEW

**Comment by davvd** [comment]
@mkordas Thanks for the pull request, let me find a reviewer..

**Comment by davvd** [comment]
@pinaf please review, thanks

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java [hunk 1]
```diff
@@ -112,6 +113,44 @@ public void understandsMethodReferences() throws Exception {
...
+                "    }",
+                "  }",
+                "}"
+            )
+        );
+        final StringWriter writer = new StringWriter();
+        Logger.getRootLogger().addAppender(
+            new WriterAppender(new SimpleLayout(), writer)
+        );
+        final Validator validator = new PMDValidator();
```
Comment (pinaf) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java]: @mkordas we can inline this variable

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java [hunk 2]
```diff
@@ -112,6 +113,44 @@ public void understandsMethodReferences() throws Exception {
...
+        validator.validate(env);
+        MatcherAssert.assertThat(
+            writer.toString(),
+            Matchers.allOf(
+                Matchers.not(Matchers.containsString("UnusedPrivateMethod")),
```
Comment (pinaf) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java]: @mkordas I believe it is `UnusedPrivateField`

**Comment by pinaf** [comment]
@mkordas 2 comments

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java (hunk 1 above)
Comment (mkordas) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java]: @pinaf thanks, variable inlined

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java (hunk 2 above)
Comment (mkordas) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java]: @pinaf good catch, name changed

**Comment by mkordas** [comment]
@pinaf fixes to the comments are in separate commit, please review

**Comment by pinaf** [comment]
@rultor merge

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java [hunk 3]
```diff
@@ -112,6 +113,43 @@ public void understandsMethodReferences() throws Exception {
...
+                "      Pattern.compile(\"hey\");",
+                "    public String doSomething() {",
+                "      return Foo.Bar.TEST.toString();",
+                "    }",
+                "  }",
+                "}"
+            )
+        );
+        final StringWriter writer = new StringWriter();
+        Logger.getRootLogger().addAppender(
```
Comment (krzyk) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java]: @mkordas you should also remove the appender after test, and `writer.flush()` will help in some cases, so it is better to add it.

**Comment by krzyk** [comment]
@mkordas one comment

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java (hunk 3 above)
Comment (mkordas) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java]: @krzyk thanks for spotting it, fixed

**Comment by mkordas** [comment]
@krzyk fix for your remark added in separate commit

File: qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java [hunk 4]
```diff
@@ -112,6 +114,47 @@ public void understandsMethodReferences() throws Exception {
...
+        new PMDValidator().validate(env);
+        writer.flush();
+        MatcherAssert.assertThat(
+            writer.toString(),
+            Matchers.allOf(
+                Matchers.not(Matchers.containsString("UnusedPrivateField")),
+                Matchers.containsString("No PMD violations found in 1 files")
+            )
+        );
+        Logger.getRootLogger().removeAppender(appender);
```
Comment (krzyk) [inline qulice-pmd/src/test/java/com/qulice/pmd/PMDValidatorTest.java]: @mkordas it would be better to have this in `finally` block

**Comment by krzyk** [comment]
@mkordas one more comment

**Comment by mkordas** [comment]
@krzyk right, we should not forget about such things, thanks. Fixed in f81211834e199a5cc0c3a7978d5ccdf77136c440

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by mkordas** [comment]
@krzyk apparently we get `/src/main/java/foo/CodeInConstructor.java[5-5]: Avoid unused private fields such as 'another'. (UnusedPrivateField)` from another test and assertion failed. What do you suggest to do here?

**Comment by mkordas** [comment]
@krzyk apparently the problem was connected to branch being out-of date. I've merged master and added new commit to fix the problem

**Comment by krzyk** [comment]
@mkordas thanks

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by davvd** [comment]
@pinaf Thanks for your contribution, **19 mins** was added to your account, payment ID is `AP-3H111895D11111827`, 67 hours and 29 mins spent total; review comments (c=4) added as a bonus; +19 added to your rating, at the moment it is: [+5125](http://www.netbout.com/b/35128?open=rating)

**Comment by davvd** [comment]
@rultor deploy

**Pull request merged**